elastica/modules/contact.py

```diff
@@ -61,30 +63,36 @@ def _finalize_contact(self) -> None:
         # to apply the contacts to
         # Technically we can use another array but it its one more book-keeping
         # step. Being lazy, I put them both in the same array
-        self._contacts[:] = [(*contact.id(), contact()) for contact in self._contacts]
-
-        # check contact order
-        for (
-            first_sys_idx,
-            second_sys_idx,
-            contact,
-        ) in self._contacts:
-            contact._check_systems_validity(
-                self._systems[first_sys_idx],
-                self._systems[second_sys_idx],
-            )
+        for contact in self._contacts:
+            first_sys_idx, second_sys_idx = contact.id()
+            contact_instance = contact.instantiate()
 
-    def _call_contacts(self, time: float):
-        for (
-            first_sys_idx,
-            second_sys_idx,
-            contact,
-        ) in self._contacts:
-            contact.apply_contact(
+            contact_instance._check_systems_validity(
                 self._systems[first_sys_idx],
                 self._systems[second_sys_idx],
             )
 
+            def apply_contact(time):
+                contact_instance.apply_contact(
+                    system_one=self._systems[first_sys_idx],
+                    system_two=self._systems[second_sys_idx],
+                )
+
+            self._feature_group_synchronize.add_operators(contact, [apply_contact])
+
+            self.warnings(contact)
```

Comment:
smart way to check. Only issue, is what happens if we have rod-rod contact and frictional plane at the same time?

Reply:
You'll get warning 😄

Reply:
Or two contacts.